deepinv/training/trainer.py

```diff
@@ -641,30 +675,44 @@ def train(
         self.setup_train()
 
         for epoch in range(self.epoch_start, self.epochs):
+            self.reset_metrics()
 
-            # clean metrics
-            self.logs_total_loss_train.reset()
-            self.logs_total_loss_eval.reset()
+            ## Training
+            self.current_iterators = [iter(loader) for loader in self.train_dataloader]
 
-            for l in self.logs_losses_train:
-                l.reset()
+            batches = min(
+                [len(loader) - loader.drop_last for loader in self.train_dataloader]
+            )
 
-            for l in self.logs_metrics_train:
-                l.reset()
+            self.model.train()
+            for i in (
+                progress_bar := tqdm(
+                    range(batches),
+                    ncols=150,
+                    disable=(not self.verbose or not self.show_progress_bar),
+                )
+            ):
+                progress_bar.set_description(f"Train epoch {epoch + 1}/{self.epochs}")
+                self.step(
+                    epoch, progress_bar, train=True, last_batch=(i == batches - 1)
+                )
 
-            for l in self.logs_metrics_eval:
-                l.reset()
+            self.loss_history.append(self.logs_total_loss_train.avg)
+
+            if self.scheduler:
+                self.scheduler.step()
 
             ## Evaluation
             perform_eval = self.eval_dataloader and (
-                (epoch + 1) % self.eval_interval == 0 or epoch + 1 == self.epochs
+                epoch % self.eval_interval == 0 or epoch == self.epochs
```

Comment:
Surely `epoch` will never `==self.epochs`? I think both should be kept as `epoch+1`

Reply:
I've changed this to always perform an eval on the first iteration (it is good to check where the model starts in my opinion)